pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -371,21 +375,30 @@ func (m *mergeQuerier) Select(ctx context.Context, sortSeries bool, hints *stora
 		jobs = append(jobs, id)
 	}
 
-	return m.innerSelectFunc(ctx, jobs, start, m.upstreamQueryWaitDuration, m.upstream, m.idLabelName, m.maxConcurrency, sortSeries, hints, filteredMatchers...)
+	wrMergeQuerier := WaitRecordingMergeQuerier{
+		start:                     start,
+		upstreamQueryWaitDuration: m.upstreamQueryWaitDuration,
+		upstream:                  m.upstream,
+	}
+
+	return m.multiTenantSelectFunc(ctx, jobs, wrMergeQuerier, m.idLabelName, m.maxConcurrency, sortSeries, hints, filteredMatchers...)
 }
 
-type InnerSelectFunc func(ctx context.Context, jobs []string, start time.Time, upstreamQueryWaitDuration prometheus.Histogram, upstream MergeQuerierUpstream, idLabelName string, maxConcurrency int, sortSeries bool, hints *storage.SelectHints, filteredMatchers ...*labels.Matcher) storage.SeriesSet
+// MultiTenantSelectFunc is an overrideable function that queries multiple tenants in parallel.
+// By default, jobs would be the list of tenant IDs, idLabelName is used to add a label to the series
+// to identify the tenant it belongs to, maxConcurrency is the maximum number of concurrent queries allowed,
+// and ctx, sortSeries, hints, and matchers are the same as the Select method.
+type MultiTenantSelectFunc func(ctx context.Context, jobs []string, wrMergeQuerier WaitRecordingMergeQuerier, idLabelName string, maxConcurrency int, sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet
 
-func defaultInnerSelectFunc(ctx context.Context, jobs []string, start time.Time, upstreamQueryWaitDuration prometheus.Histogram, upstream MergeQuerierUpstream, idLabelName string, maxConcurrency int, sortSeries bool, hints *storage.SelectHints, filteredMatchers ...*labels.Matcher) storage.SeriesSet {
+func defaultMultiTenantSelectFunc(ctx context.Context, jobs []string, wrMergeQuerier WaitRecordingMergeQuerier, idLabelName string, maxConcurrency int, sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
```

Comment:
shouldn't we still pass `MergeQuerierUpstream`? `WaitRecordingMergeQuerier` should be just an implementation of that interface